domain/autocert/service/service.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+	"golang.org/x/crypto/acme/autocert"
+)
+
+// State describes retrieval and persistence methods for storage.
+type State interface {
+	// Put implements autocert.Cache.Put.
+	Put(ctx context.Context, name string, data []byte) error
+
+	// Get implements autocert.Cache.Get.
+	Get(ctx context.Context, name string) ([]byte, error)
+
+	// Delete implements autocert.Cache.Delete.
+	Delete(ctx context.Context, name string) error
+}
+
+// Logger facilitates emitting log messages.
+type Logger interface {
+	Tracef(string, ...interface{})
+}
+
+// Service provides the API for working with autocert cache. This service
+// implements autocert.Cache interface.
+type Service struct {
+	st     State
+	logger Logger
+}
+
+// NewService returns a new service reference wrapping the input state.
+func NewService(st State, logger Logger) *Service {
+	return &Service{
+		st:     st,
+		logger: logger,
+	}
+}
+
+// Put implements autocert.Cache.Put.
+func (s *Service) Put(ctx context.Context, name string, data []byte) error {
+	s.logger.Tracef("storing autocert %s with contents '%+v' in the autocert cache", name, string(data))
+	return s.st.Put(ctx, name, data)
+}
+
+// Get implements autocert.Cache.Get.
+func (s *Service) Get(ctx context.Context, name string) ([]byte, error) {
```

Comment:
This style was mentioned in London. I am of the opinion that we should not conflate domain services with single entities, even though they might be in some cases.

So `entity.Get` makes sense, but `service.Get` is ambiguous. I would prefer to be explicit with naming. Will socialise this.

Reply:
I 100% agree with you on the general case, but on this particular case the Service must satisfy the interface `autocert.Cache` (third party lib).